### PR TITLE
[neox-2.x] root p2p message handle

### PR DESCRIPTION
**Pull request by KickSeason**
### Step 6/7  from `neo-ngd/neox` to `neox-2.x`
changes in neo/Network
* root messages handle
* root related rpc

@erikzhang @shargon  Can you please review this one?

### CODE REVIEW

File: neo/Network/RPC/RpcServer.cs [hunk 1]
```diff
@@ -749,5 +772,86 @@ private JObject ValidateAddress(string address)
             json["isvalid"] = scriptHash != null;
             return json;
         }
+
+        private JObject GetStateHeight()
+        {
+            var json = new JObject();
+            json["blockheight"] = Blockchain.Singleton.Height;
```
Comment (shargon) [inline neo/Network/RPC/RpcServer.cs:779]: camelcase or _

File: neo/Network/P2P/ProtocolHandler.cs [hunk 2]
```diff
@@ -275,6 +311,12 @@ private void OnHeadersMessageReceived(HeadersPayload payload)
 
         private void OnInventoryReceived(IInventory inventory)
         {
+            if (inventory.InventoryType == InventoryType.StateRoot && !knownHashes.Contains(inventory.Hash))
```
Comment (shargon) [inline neo/Network/P2P/ProtocolHandler.cs:314]: if it is in known hashes, it should not continue and is sent to the task manager.

File: neo/Network/P2P/ProtocolHandler.cs [hunk 3]
```diff
@@ -155,6 +164,33 @@ private void OnAddrMessageReceived(AddrPayload payload)
...
+            var state_roots = new List<StateRoot>();
+            for (uint i = 0; i < count; i++)
+            {
+                var state = Blockchain.Singleton.GetStateRoot(start + i);
+                if (state?.Flag == StateRootVerifyFlag.Verified)
+                {
+                    state_roots.Add(state.StateRoot);
+                    continue;
+                }
+                break;
```
Comment (shargon) [inline neo/Network/P2P/ProtocolHandler.cs:180]: if only one it's not verified, we should continue?

File: neo/Network/RPC/RpcServer.cs [hunk 4]
```diff
@@ -749,5 +772,86 @@ private JObject ValidateAddress(string address)
...
+        private JObject VerifyProof(UInt256 state_root, byte[] proof_bytes)
+        {
+            var json = new JObject();
+            var proof = new HashSet<byte[]>(ByteArrayEqualityComparer.Default);
+            byte[] key;
+            using (MemoryStream ms = new MemoryStream(proof_bytes, false))
+            using (BinaryReader reader = new BinaryReader(ms, Encoding.UTF8))
+            {
+                key = reader.ReadVarBytes();
+                var count = reader.ReadVarInt();
```
Comment (shargon) [inline neo/Network/RPC/RpcServer.cs:839]: MaxValue

File: neo/Network/P2P/ProtocolHandler.cs [hunk 5]
```diff
@@ -155,6 +164,30 @@ private void OnAddrMessageReceived(AddrPayload payload)
             });
         }
 
+        private void OnGetStateRootsReceived(GetStateRootsPayload payload)
+        {
+            var start = payload.StartIndex;
```
Comment (shargon) [inline neo/Network/P2P/ProtocolHandler.cs:169]: We don't need this variable

File: neo/Network/P2P/ProtocolHandler.cs [hunk 6]
```diff
@@ -94,6 +94,15 @@ private void OnMessage(Message msg)
                 case "consensus":
                     OnInventoryReceived(msg.GetPayload<ConsensusPayload>());
                     break;
+                case "root":
+                    OnInventoryReceived(msg.GetPayload<StateRoot>());
```
Comment (shargon) [inline neo/Network/P2P/ProtocolHandler.cs:98]: Where it's send this message?

Comment (KickSeason) [reply neo/Network/P2P/ProtocolHandler.cs:98]: https://github.com/neo-ngd/neo/blob/99061343bf91b34a8264d32b684fc92f2d343ee7/neo/Consensus/ConsensusService.cs#L123

https://github.com/neo-project/neo/blob/24ecde1ea3942540778d7e226ddfc69407751060/neo/Ledger/BlockChain.State.cs#L64

**Comment by Tommo-L** [comment]
Let's merge?

**Pull request merged**